warp/yul/DeadcodeEliminator.py

```diff
@@ -0,0 +1,172 @@
+        )
+        live_vars -= fset({arg.name for arg in node.parameters})
+        return (
+            live_vars,
+            ast.FunctionDefinition(
+                name=node.name,
+                parameters=node.parameters,
+                return_variables=node.return_variables,
+                body=body,
+            ),
+        )
+
+    def visit_if(
+        self, node: ast.If, live_vars: Set[str] = None, return_vars: Set[str] = None
+    ) -> Tuple[LiveVars, ast.If]:
+        if any(
+            isinstance(statement, ast.Leave)
+            for statement in node.body.statements
+            + (node.else_body.statements if node.else_body else ())
+        ):
+            live_vars |= return_vars
+
+        (live_vars_cond, cond) = self.visit(node.condition)
+        (live_vars_body, body) = self.visit_block(node.body, live_vars)
+        (live_vars_else, else_body) = (
+            self.visit_block(node.else_body, live_vars)
+            if node.else_body
+            else (fset(), None)
+        )
+        return (
+            live_vars_cond | live_vars_body | live_vars_else,
+            ast.If(condition=cond, body=body, else_body=else_body),
+        )
+
+    def visit_switch(self, node: ast.Switch) -> Tuple[LiveVars, ast.Switch]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no switches, run SwitchToIfVisitor first"
+
+    def visit_forloop(self, node: ast.ForLoop) -> Tuple[LiveVars, ast.ForLoop]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no for loops, run ForLoopIliminator first"
+
+    def visit_break(self, node: ast.Break) -> Tuple[LiveVars, ast.Break]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no breaks, run ForLoopIliminator first"
+        return (fset(), node)
+
+    def visit_continue(self, node: ast.Continue) -> Tuple[LiveVars, ast.Continue]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no continues, run ForLoopIliminator first"
+        return (fset(), node)
+
+    def visit_leave(self, node: ast.Leave) -> Tuple[LiveVars, ast.Leave]:
+        return (fset(), node)
+
+    def visit_block(
+        self, node: ast.Block, live_vars: Set[str] = None, return_vars: Set[str] = None
+    ) -> Tuple[LiveVars, ast.Block]:
+        if not node.statements:
+            return (fset(), node)
+        if live_vars is None:
+            live_vars = fset()
+        if return_vars is not None:
+            live_vars |= return_vars
+        statements = []
+        for n in reversed(node.statements):
+            keep = True
+            if isinstance(n, ast.If):
+                (current_live_vars, statement) = self.visit(n, live_vars, return_vars)
+                live_vars = current_live_vars
```

Comment:
At first I though that the set of live vars shrinks here, but apparently it's extended because we pass pre-existing `live_vars` to `self.visit` right? Is there any reason that we should pass it as an argument? 
I in general don't understand why if's treatment differs from other nodes, could you please explain?
Ah, I see that is because of passing `return_vars`. If we use a contextmanager as I suggested in the other commen, maybe we can get rid of special treatment for if?

Reply:
Actually there is a subtle bug in how I've done this. I wanted to kill any variable (declared before the scope of an if) that was assigned to in both branches of the if statement. However I didn't differentiate between variable declaration and assignment so any variables which are declared in both branches will kill a variable declaration of the same name before. I'm going to modify this to take away the `special treatment` which will underaproximate the analysis. I don't think we hit that case often though with the current ssa.